### ISSUE TITLE
Crash building glob.c from gnulib in purecap

### ISSUE THREAD

**Issue by bsdjhb** [CTSRD-CHERI/llvm-project#337]
Trying to build a purecap GDB using cheri clang I got the following crash:

```
  CC       glob.o
Assertion failed: (VT.isInteger() && Operand.getValueType().isInteger() && "Invalid SIGN_EXTEND!"), function getNode, file /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp, line 4538.
 #0 0x0000000804360178 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/john/work/cheri/git/llvm/llvm/lib/Support/Unix/Signals.inc:494:13
 #1 0x0000000804360556 __assert /home/john/work/cheri/git/llvm/llvm/lib/Support/Unix/Signals.inc:582:3
 #2 0x0000000805dad3e9 llvm::SelectionDAG::getNode(unsigned int, llvm::SDLoc const&, llvm::EVT, llvm::SDValue, llvm::SDNodeFlags) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:0:5
 #3 0x0000000805db268d llvm::SelectionDAG::getSExtOrTrunc(llvm::SDValue, llvm::SDLoc const&, llvm::EVT) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:1123:3
 #4 0x0000000805d957a5 llvm::SDNode::getValueType(unsigned int) const /home/john/work/cheri/git/llvm/llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1004:5
 #5 0x0000000805d957a5 llvm::SDValue::getValueType() const /home/john/work/cheri/git/llvm/llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1152:0
 #6 0x0000000805d957a5 llvm::SelectionDAGBuilder::visitMemPCpyCall(llvm::CallInst const&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7397:0
 #7 0x0000000805d6b8b8 llvm::SelectionDAGBuilder::visitCall(llvm::CallInst const&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7661:13
 #8 0x0000000805d5f34e llvm::Value::getValueID() const /home/john/work/cheri/git/llvm/llvm/include/llvm/IR/Value.h:464:12
 #9 0x0000000805d5f34e llvm::isa_impl<llvm::Instruction, llvm::Value, void>::doit(llvm::Value const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/IR/Value.h:802:0
#10 0x0000000805d5f34e llvm::isa_impl_cl<llvm::Instruction, llvm::Value const*>::doit(llvm::Value const*) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:106:0
#11 0x0000000805d5f34e llvm::isa_impl_wrap<llvm::Instruction, llvm::Value const*, llvm::Value const*>::doit(llvm::Value const* const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:132:0
#12 0x0000000805d5f34e llvm::isa_impl_wrap<llvm::Instruction, llvm::Value const* const, llvm::Value const*>::doit(llvm::Value const* const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:122:0
#13 0x0000000805d5f34e bool llvm::isa<llvm::Instruction, llvm::Value const*>(llvm::Value const* const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:142:0
#14 0x0000000805d5f34e llvm::cast_retty<llvm::Instruction, llvm::Value const*>::ret_type llvm::dyn_cast<llvm::Instruction, llvm::Value const>(llvm::Value const*) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:343:0
#15 0x0000000805d5f34e llvm::FPMathOperator::classof(llvm::Value const*) /home/john/work/cheri/git/llvm/llvm/include/llvm/IR/Operator.h:374:0
#16 0x0000000805d5f34e llvm::isa_impl<llvm::FPMathOperator, llvm::Instruction, void>::doit(llvm::Instruction const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:58:0
#17 0x0000000805d5f34e llvm::isa_impl_cl<llvm::FPMathOperator, llvm::Instruction const*>::doit(llvm::Instruction const*) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:106:0
#18 0x0000000805d5f34e llvm::isa_impl_wrap<llvm::FPMathOperator, llvm::Instruction const*, llvm::Instruction const*>::doit(llvm::Instruction const* const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:132:0
#19 0x0000000805d5f34e llvm::isa_impl_wrap<llvm::FPMathOperator, llvm::Instruction const* const, llvm::Instruction const*>::doit(llvm::Instruction const* const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:122:0
#20 0x0000000805d5f34e bool llvm::isa<llvm::FPMathOperator, llvm::Instruction const*>(llvm::Instruction const* const&) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:142:0
#21 0x0000000805d5f34e llvm::cast_retty<llvm::FPMathOperator, llvm::Instruction const*>::ret_type llvm::dyn_cast<llvm::FPMathOperator, llvm::Instruction const>(llvm::Instruction const*) /home/john/work/cheri/git/llvm/llvm/include/llvm/Support/Casting.h:343:0
#22 0x0000000805d5f34e llvm::SelectionDAGBuilder::visit(llvm::Instruction const&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:1097:0
#23 0x0000000805df98cd llvm::ilist_node_base<true>::getNext() const /home/john/work/cheri/git/llvm/llvm/include/llvm/ADT/ilist_node_base.h:43:45
#24 0x0000000805df98cd llvm::ilist_node_impl<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void> >::getNext() const /home/john/work/cheri/git/llvm/llvm/include/llvm/ADT/ilist_node.h:74:0
#25 0x0000000805df98cd llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>::operator++() /home/john/work/cheri/git/llvm/llvm/include/llvm/ADT/ilist_iterator.h:158:0
#26 0x0000000805df98cd llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>, llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void>, false, true>, bool&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:694:0
#27 0x0000000805df910c llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1595:11
#28 0x0000000805df6175 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:501:7
#29 0x0000000801b90dd0 llvm::MipsDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/john/work/cheri/git/llvm/llvm/lib/Target/Mips/MipsISelDAGToDAG.cpp:58:32
#30 0x00000008034fcb5b llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/john/work/cheri/git/llvm/llvm/lib/CodeGen/MachineFunctionPass.cpp:73:13
#31 0x00000008039b529d llvm::FPPassManager::runOnFunction(llvm::Function&) /home/john/work/cheri/git/llvm/llvm/lib/IR/LegacyPassManager.cpp:1650:27
#32 0x00000008039b55c3 llvm::FPPassManager::runOnModule(llvm::Module&) /home/john/work/cheri/git/llvm/llvm/lib/IR/LegacyPassManager.cpp:1687:13
#33 0x00000008039b5ade (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/john/work/cheri/git/llvm/llvm/lib/IR/LegacyPassManager.cpp:1754:27
#34 0x00000008039b5ade llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/john/work/cheri/git/llvm/llvm/lib/IR/LegacyPassManager.cpp:1867:0
#35 0x0000000804e026fe (anonymous namespace)::EmitAssemblyHelper::EmitAssembly(clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) /home/john/work/cheri/git/llvm/clang/lib/CodeGen/BackendUtil.cpp:929:3
#36 0x0000000804e026fe clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::DataLayout const&, llvm::Module*, clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) /home/john/work/cheri/git/llvm/clang/lib/CodeGen/BackendUtil.cpp:1537:0
```

creduce got the source down to this:

```
#if !0
#if 3
#define a(b, c, d) c b d
#endif
typedef __SIZE_TYPE__ e;
#if 3
a(mempcpy, void *, (void *, *, e));
#endif
#endif
f, g;
h() { mempcpy(h, &f, g); }
```

The original command line:

```
# Original command:  "/usr/cheri/bin/clang-8" "-cc1" "-triple" "mips64c128-unknown-freebsd13.0-purecap" "-emit-obj" "-disable-free" "-main-file-name" "glob.c" "-mrelocation-model" "pic" "-pic-level" "1" "-mthread-model" "posix" "-mdisable-fp-elim" "-relaxed-aliasing" "-masm-verbose" "-mconstructor-aliases" "-fuse-init-array" "-target-cpu" "mips4" "-target-feature" "-noabicalls" "-target-feature" "+soft-float" "-target-feature" "+chericap" "-target-feature" "+cheri128" "-target-abi" "purecap" "-Wmips-cheri-prototypes" "-msoft-float" "-mfloat-abi" "soft" "-cheri-size" "128" "-mllvm" "-cheri-cap-table-abi=pcrel" "-mllvm" "-mxcaptable=false" "-dwarf-column-info" "-debug-info-kind=standalone" "-dwarf-version=2" "-debugger-tuning=gdb" "-coverage-notes-file" "/home/john/work/cheri/git/gdb/obj.cheriabi128/gdb/build-gnulib/import/glob.gcno" "-resource-dir" "/usr/cheri/lib/clang/9.0.0" "-dependency-file" ".deps/glob.Tpo" "-sys-header-deps" "-MP" "-MT" "glob.o" "-D" "HAVE_CONFIG_H" "-I" "." "-I" "../../../../gdb/gnulib/import" "-I" ".." "-D" "RL_NO_COMPAT" "-D" "LIBICONV_PLUG" "-isysroot" "/home/john/work/cheri/rootfs128" "-O2" "-Wno-unused-function" "-Wno-unused-variable" "-Wno-absolute-value" "-Wno-parentheses-equality" "-Wno-unknown-warning-option" "-fdebug-compilation-dir" "/home/john/work/cheri/git/gdb/obj.cheriabi128/gdb/build-gnulib/import" "-ferror-limit" "19" "-fmessage-length" "80" "-mstack-alignment=16" "-fobjc-runtime=gnustep" "-fdiagnostics-show-option" "-fcolor-diagnostics" "-vectorize-loops" "-vectorize-slp" "-cheri-linker" "-faddrsig" "-o" "glob.o" "-x" "c" "../../../../gdb/gnulib/import/glob.c"
 "/usr/cheri/bin/clang-8" "-cc1" "-triple" "mips64c128-unknown-freebsd13.0-purecap" "-emit-obj" "-disable-free" "-main-file-name" "glob.c" "-mrelocation-model" "pic" "-pic-level" "1" "-mthread-model" "posix" "-mdisable-fp-elim" "-relaxed-aliasing" "-masm-verbose" "-mconstructor-aliases" "-fuse-init-array" "-target-cpu" "mips4" "-target-feature" "-noabicalls" "-target-feature" "+soft-float" "-target-feature" "+chericap" "-target-feature" "+cheri128" "-target-abi" "purecap" "-Wmips-cheri-prototypes" "-msoft-float" "-mfloat-abi" "soft" "-cheri-size" "128" "-mllvm" "-cheri-cap-table-abi=pcrel" "-mllvm" "-mxcaptable=false" "-dwarf-column-info" "-debug-info-kind=standalone" "-dwarf-version=2" "-debugger-tuning=gdb" "-coverage-notes-file" "/home/john/work/cheri/git/gdb/obj.cheriabi128/gdb/build-gnulib/import/glob.gcno" "-sys-header-deps" "-D" "HAVE_CONFIG_H" "-D" "RL_NO_COMPAT" "-D" "LIBICONV_PLUG" "-O2" "-Wno-unused-function" "-Wno-unused-variable" "-Wno-absolute-value" "-Wno-parentheses-equality" "-Wno-unknown-warning-option" "-ferror-limit" "19" "-fmessage-length" "80" "-mstack-alignment=16" "-fobjc-runtime=gnustep" "-fdiagnostics-show-option" "-fcolor-diagnostics" "-vectorize-loops" "-vectorize-slp" "-cheri-linker" "-faddrsig" "-x" "c" "glob-bd94cf.c"
```

**Comment by jrtc27** [comment]
Fix should be something like (hand-written patch, expect bugs):

```diff
--- a/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp
@@ -7391,11 +7391,12 @@
   DAG.setRoot(MC);
 
   // Check if Size needs to be truncated or extended.
-  Size = DAG.getSExtOrTrunc(Size, sdl, Dst.getValueType());
+  unsigned DstAS = I.getArgOperand(0)->getType()->getPointerAddressSpace();
+  EVT OffsetVT = MVT::getIntegerVT(DAG.getDataLayout().getIndexSizeInBits(DstAS));
+  Size = DAG.getSExtOrTrunc(Size, sdl, OffsetVT);
 
   // Adjust return pointer to point just past the last dst byte.
-  SDValue DstPlusSize = DAG.getNode(ISD::ADD, sdl, Dst.getValueType(),
-                                    Dst, Size);
+  SDValue DstPlusSize = DAG.getPointerAdd(sdl, Dst, Size);
   setValue(&I, DstPlusSize);
   return true;
 }
```

**Comment by bsdjhb** [comment]
That patch "worked" for me in that I was able to compile all of gdb.  I haven't yet run it to see if it works ok.

**Comment by arichardson** [comment]
@jrtc27 Did you commit this change?

**Comment by jrtc27** [comment]
Yes, fixed in bc43903356a1f2f6a75c0913c8aff3c52614638b but neglected to close this.
